### PR TITLE
[SPARK-12498][SQL][MINOR] BooleanSimplication simplification

### PR DESCRIPTION
**Pull request by liancheng**
Scala syntax allows binary case classes to be used as infix operator in pattern matching. This PR makes use of this syntax sugar to make `BooleanSimplification` more readable.

### CODE REVIEW

**Comment by cloud-fan** [comment]
It's much more readable! LGTM

**Comment by rxin** [comment]
It seems to me the change is not necessary and might actually complicate the analyzer code itself due to the reliance on the dsl implicits and this "unapply" (I didn't even know it existed.....)

One thing I do like though is the True and False, although I'd rename it TrueLiteral and FalseLiteral, in order to make it clear it is a literal.

**Comment by rxin** [comment]
Update: the unapply is probably fine. I still wouldn't have the dsl though.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -473,112 +474,97 @@ object OptimizeIn extends Rule[LogicalPlan] {
...
-              // (a && b) || (a && b && c && ...) => a && b
-              common.reduce(And)
-            } else {
-              // (a && b && c && ...) || (a && b && d && ...) =>
-              // ((c && ...) || (d && ...)) && a && b
-              (common :+ Or(ldiff.reduce(And), rdiff.reduce(And))).reduce(And)
-            }
+            // (a && b && c && ...) || (a && b && d && ...) =>
+            // ((c && ...) || (d && ...)) && a && b
+            (common :+ Or(ldiff.reduce(And), rdiff.reduce(And))).reduce(And)
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: (Code for common factor elimination for conjunction and disjunction is untouched except for indentation changes.)

**Comment by liancheng** [comment]
Thanks for the review! I reverted the DSL part and those changes that are not obviously correct.

**Comment by rxin** [comment]
I'm going to merge this. Thanks.

**Pull request closed without merging**